src/broadcast/spectateWorker.ts

```diff
@@ -0,0 +1,82 @@
+// NOTE: This module cannot use electron-log, since it for some reason
+// fails to obtain the paths required for file transport to work
+// when in Node worker context.
+
+import { ReplayCommunication } from "@dolphin/types";
+import { ModuleMethods } from "threads/dist/types/master";
+import { Observable, Subject } from "threads/observable";
+import { expose } from "threads/worker";
+
+import { SpectateManager } from "./spectateManager";
+import { BroadcasterItem, SpectateEvent } from "./types";
+
+export interface Methods {
+  destroyWorker: () => Promise<void>;
+  startSpectate(broadcastId: string, targetPath: string): Promise<void>;
+  stopSpectate(broadcastId: string): Promise<void>;
+  dolphinClosed(playbackId: string): Promise<void>;
+  refreshBroadcastList(authToken: string): Promise<void>;
+  getLogObservable(): Observable<string>;
+  getErrorObservable(): Observable<string>;
+  getBroadcastListObservable(): Observable<BroadcasterItem[]>;
+  getSpectateDetailsObservable(): Observable<{ playbackId: string; replayComm: ReplayCommunication }>;
+}
+
+export type WorkerSpec = ModuleMethods & Methods;
+
+const spectateManager = new SpectateManager();
+
+const logSubject = new Subject<string>();
+const errorSubject = new Subject<string>();
+const broadcastListSubject = new Subject<BroadcasterItem[]>();
+const spectateDetailsSubject = new Subject<{ playbackId: string; replayComm: ReplayCommunication }>();
+
+// Forward the events to the renderer
+spectateManager.on(SpectateEvent.BROADCAST_LIST_UPDATE, async (data: BroadcasterItem[]) => {
+  broadcastListSubject.next(data);
+});
+
+spectateManager.on(SpectateEvent.LOG, async (msg: string) => {
+  logSubject.next(msg);
+});
+
+spectateManager.on(SpectateEvent.ERROR, async (errorMsg: string) => {
+  errorSubject.next(errorMsg);
+});
+
+spectateManager.on(SpectateEvent.PLAY_FILE, async (playbackId: string, replayComm: ReplayCommunication) => {
+  spectateDetailsSubject.next({ playbackId, replayComm });
+});
```

Comment:
I'm not really a fan of spectate worker emitting a "play file" event. It shouldn't really need to know about playing things back. I think that should be main's job. The event should probably be called `NEW_GAME` or `NEW_FILE` or `GAME_START` or something which should return an Observable string (the SLP filename). then it should be main's job to generate the playback ID and replay communication.

Reply:
Sure